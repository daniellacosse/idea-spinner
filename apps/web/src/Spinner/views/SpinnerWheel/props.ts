export interface SpinnerWheelProps {
  description?: string;
  isSpinning: boolean;
  name: string;
  tickDuration: string;
  value?: string;
}

export default {
  description:
    "Clowns are a curious breed who sustain themselves " +
    "on the laugher of children.",
  isSpinning: false,
  name: "Demo Wheel",
  tickDuration: "200ms",
  value: "Clowns"
} as SpinnerWheelProps;
