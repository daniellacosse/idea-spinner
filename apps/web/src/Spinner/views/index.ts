export * from "./SpinnerWheel";
export { default as Spinner } from "./Spinner.vue";
export { SpinnerProps } from "./props";
