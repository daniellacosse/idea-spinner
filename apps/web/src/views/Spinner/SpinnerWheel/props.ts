export interface SpinnerWheelProps {
  isSpinning: boolean;
  name: string;
  value: string;
}

export default {
  isSpinning: false,
  name: "Demo Wheel",
  value: "Clowns"
} as SpinnerWheelProps;
