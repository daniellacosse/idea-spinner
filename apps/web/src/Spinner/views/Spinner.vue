<script lang="ts" setup>
import { ref } from "vue";
import { SpinnerWheel, SpinnerWheelProps } from "./SpinnerWheel";

// see: https://github.com/vuejs/core/issues/4294
interface SpinnerProps {
  hasSpun: boolean;
  isSpinning: boolean;
  maxColumns?: number;
  wheels: SpinnerWheelProps[];
}

const props = withDefaults(defineProps<SpinnerProps>(), { maxColumns: 4 });

const rowCount = ref(Math.ceil(props.wheels.length / props.maxColumns));

defineEmits(["spin"]);
</script>

<template>
  <article class="Spinner">
    <ul :class="{
      'SpinnerWheels': true,
      'SpinnerWheels--waiting': !hasSpun
    }">
      <SpinnerWheel v-for="wheel in wheels" :key="wheel.name" v-bind="wheel" />
    </ul>

    <button class="SpinnerLever" @click.prevent="$emit('spin', true)"
      :disabled="isSpinning">
      {{ isSpinning ? "Spinning..." : "Click to Spin!" }}
    </button>
  </article>
</template>

<style scoped>
* {
  all: initial;
}

.Spinner {
  align-items: center;
  display: flex;
  flex-direction: column;
  gap: var(--gutter-standard);
  height: 100%;
  max-width: var(--max-supported-width);
  width: 100%;
}

.SpinnerWheels {
  display: grid;
  flex-grow: 1;
  grid-auto-flow: column;
  grid-gap: var(--gutter-standard);
  grid-template-rows: repeat(v-bind(rowCount), auto);
  width: 100%;
  opacity: 1;
  transition: opacity var(--animation-timing) ease-out;
}

.SpinnerWheels--waiting {
  opacity: 0.5;
}

.SpinnerLever {
  background: var(--color-primary-variant);
  border-radius: calc(var(--gutter-standard) / 2);
  color: var(--color-primary-variant-foreground);
  cursor: pointer;
  font-size: var(--gutter-wide);
  padding: var(--gutter-narrow) var(--gutter-wide);
  font-family: var(--font-primary);
  text-align: center;
}

.SpinnerLever[disabled] {
  background: var(--color-primary-foreground);
  cursor: not-allowed;
}

@media screen and (max-width: 760px) {
  .SpinnerWheels {
    grid-auto-flow: row;
  }
}
</style>
