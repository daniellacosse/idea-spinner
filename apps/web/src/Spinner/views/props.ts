import { SpinnerWheelProps } from "./SpinnerWheel";

export interface SpinnerProps {
  hasSpun: boolean;
  isSpinning: boolean;
  maxColumns?: number;
  wheels: SpinnerWheelProps[];
}

export default {
  hasSpun: false,
  isSpinning: false,
  maxColumns: 4,
  wheels: [{
    isSpinning: true,
    name: "color",
    value: "red",
  }, {
    isSpinning: true,
    name: "shape",
    value: "round",
  }, {
    isSpinning: false,
    name: "feeling",
    value: "sad",
  }]
} as SpinnerProps;
