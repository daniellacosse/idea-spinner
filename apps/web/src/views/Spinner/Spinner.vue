<script lang="ts" setup>
import { ref } from "vue";
import { SpinnerWheel, SpinnerWheelProps } from "./SpinnerWheel";

// see: https://github.com/vuejs/core/issues/4294
interface SpinnerProps {
  isSpinning: boolean;
  maxColumns: number;
  wheels: SpinnerWheelProps[];
}

const props = withDefaults(defineProps<SpinnerProps>(), { maxColumns: 4 });

const rowCount = ref(Math.ceil(props.wheels.length / props.maxColumns));
const leverText = ref(props.isSpinning ? "Spinning..." : "Click to Spin!");

defineEmits(["spin"]);
</script>

<template>
  <article class="Spinner">
    <ul class="SpinnerWheels">
      <SpinnerWheel v-for="wheel in wheels" :key="wheel.name" v-bind="wheel" />
    </ul>

    <button class="SpinnerLever" @click.prevent="$emit('spin', true)"
      :disabled="isSpinning">
      {{ leverText }}
    </button>
  </article>
</template>

<style scoped>
* {
  all: initial;
}

.Spinner {
  align-items: center;
  display: flex;
  flex-direction: column;
  gap: var(--gutter-standard);
  height: 100%;
  max-width: var(--max-supported-width);
  min-width: var(--min-supported-width);
  width: 100%;
}

.SpinnerWheels {
  display: grid;
  flex-grow: 1;
  grid-auto-flow: column;
  grid-gap: var(--gutter-standard);
  grid-template-rows: repeat(v-bind(rowCount), auto);
  width: 100%;
}

.SpinnerLever {
  background: var(--color-primary-variant);
  border-radius: calc(var(--gutter-standard) / 2);
  color: var(--color-primary-variant-foreground);
  cursor: pointer;
  font-size: var(--gutter-wide);
  padding: var(--gutter-narrow) var(--gutter-wide);
  text-align: center;
}

.SpinnerLever[disabled] {
  background: var(--color-primary-foreground);
  cursor: not-allowed;
}
</style>
