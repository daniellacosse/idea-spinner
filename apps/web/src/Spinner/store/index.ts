import { createStore } from "vuex";
import { SpinnerWheelProps } from "../views/SpinnerWheel";
import { Spinner, SpinnerPhysics } from "@idea-spinner/spinner";

export interface SpinnerStoreState {
  display: SpinnerWheelProps[];
  isSpinning: boolean;
  spinner?: Spinner;
  currentSpinID?: string;
}

export enum SpinnerStoreMutations {
  INITIALIZE = "spinner/initialize",
  SPIN = "spinner/spin",
  ADVANCE = "spinner/advance"
}

export const SpinnerStore = createStore<SpinnerStoreState>({
  mutations: {
    [SpinnerStoreMutations.INITIALIZE](state, wheels: Map<string, string[]>) {
      state.spinner = new Spinner({ wheels });
      state.display = [...wheels.keys()].map(
        name => ({
          isSpinning: false,
          name,
          value: "--",
        })
      );
    },
    [SpinnerStoreMutations.SPIN](state, physics: SpinnerPhysics) {
      if (!state.spinner) {
        throw new TypeError(
          "You must INITIALIZE the spinner store before starting a spin!"
        );
      }

      state.currentSpinID = state.spinner.createSpin(physics);
      state.isSpinning = true;
    },
    [SpinnerStoreMutations.ADVANCE](state, time: number) {
      if (!state.spinner) {
        throw new TypeError(
          "You must INITIALIZE the spinner and then " +
          "SPIN it before ADVANCEing that spin!"
        );
      }

      if (state.currentSpinID === undefined) {
        throw new TypeError(
          "You must SPIN the spinner store before ADVANCEing the spin!"
        );
      }

      const spinObject = state.spinner.getSpin(state.currentSpinID);

      state.spinner.advanceSpin(state.currentSpinID, time);

      if (!spinObject) {
        return;
      }

      state.display = [...spinObject.wheels.entries()].map(
        ([name, { value, isSpinning }]) => ({ name, value, isSpinning })
      );
      state.isSpinning = spinObject.isSpinning;
    }
  },
  state() {
    return {
      isSpinning: false,
      display: []
    };
  }
});
