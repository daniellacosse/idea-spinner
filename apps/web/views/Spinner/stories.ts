import Spinner from "./Spinner.vue";
import args, { SpinnerProps } from "./props";

export default {
  args,
  argTypes: {
    handleSpin: {
      action: "spin"
    }
  },
  component: Spinner,
  title: "Spinner"
};

export const Example = ({
  isSpinning,
  maxColumns,
  wheels,
  handleSpin
}: SpinnerProps & { handleSpin: () => void; }) => ({
  components: { Spinner },
  setup() {
    return { handleSpin, isSpinning, maxColumns, wheels };
  },
  template: `
    <Spinner 
      :wheels="wheels" 
      :is-spinning="isSpinning"
      :max-columns="maxColumns" 
      @spin='handleSpin'
    />
  `
});
