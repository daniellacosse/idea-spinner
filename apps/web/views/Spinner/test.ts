import props from "./props";
import { shallowMount } from "@vue/test-utils";
import Spinner from "./Spinner.vue";

test.concurrent("spinner - snapshot", async () => {
  const { element } = shallowMount(Spinner, { props });

  expect(element).toMatchSnapshot();
});

test.concurrent("spinner - spinning", async () => {
  const { element } = shallowMount(Spinner, {
    props: { ...props, isSpinning: true }
  });

  expect(element).toMatchSnapshot();
});
