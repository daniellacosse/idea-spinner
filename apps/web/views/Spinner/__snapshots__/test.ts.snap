// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<article
  class="Spinner"
>
  <ul
    class="SpinnerWheels"
  >
    
    <spinner-wheel-stub
      isspinning="true"
      name="color"
      value="red"
    />
    <spinner-wheel-stub
      isspinning="true"
      name="shape"
      value="round"
    />
    <spinner-wheel-stub
      isspinning="false"
      name="feeling"
      value="sad"
    />
    
  </ul>
  <button
    class="SpinnerLever"
  >
    Click to Spin!
  </button>
</article>
`;

exports[` 2`] = `
<article
  class="Spinner"
>
  <ul
    class="SpinnerWheels"
  >
    
    <spinner-wheel-stub
      isspinning="true"
      name="color"
      value="red"
    />
    <spinner-wheel-stub
      isspinning="true"
      name="shape"
      value="round"
    />
    <spinner-wheel-stub
      isspinning="false"
      name="feeling"
      value="sad"
    />
    
  </ul>
  <button
    class="SpinnerLever"
    disabled=""
  >
    Spinning...
  </button>
</article>
`;
